### DISCUSSION TITLE
Frage bezüglich Programmierung

### DISCUSSION THREAD

**Discussion by Manos1966** [skippermeister/OpenDTU-OnBattery#10] [Q&A] [answered]
Hallo @skippermeister 
da du eindeutig viel mehr Programmierkenntnisse hast als ich, wollte ich dich fragen:
Wir wissen, dass z.Zeit der Huawei AC charger ist "Hard-coded" an SPI Bus 2.
Das gleiche ist leider der Fall mit CMT2300A.

Wie "schwierig/aufwendig" wäre es, eins (Huawei oder CMT2300) auf SPI Bus 3 umzustellen? 🙊

**Comment by skippermeister** [comment]
Hallo @Manos1966 
Ich würde einen SPIPortmanager realisieren, ähnlich dem SerialPortmanager. Dann wäre die Zuordnung der SPIs dynamisch den jeweiligen Geräten zugeordnet.

Den Aufwand sehe ich als machbar an. Ich kann dass auch in meinem Fork programmieren, allerdings habe ich weder den CMT2300A noch das Huawei Netzteil. Ich kann das nur grundsätzlich entwickeln und in meiner Umgebung (NRF24, MCP2515 als CAN Bus alternativ zu dem internen CAN0 Bus des ESP32 mit Meanwell Netzteil). damit hätte ich 2 SPI Geräte.

Die Programmierung müsste dann jemand in die originale OpenDTU-onBattery überführen. In der original Version Dort möchte ich eigentlich noch zusätzlich programmieren.

**Reply by skippermeister** [reply]
Hallo @Manos1966 
ich habe in meinen Fork einen SPIPortManager ähnlich dem SerialPortManager realisiert. Die zwei zur Verfügung stehenden SPIs werden nun den konfigurierten Geräten dynamisch zugewiesen. Aktuell werden diese nur zugewiesen, bei Konfigurationsänderungen werden noch nicht die allocierten SPI Resourcen wieder freigegeben, so wie es beim SerialPortManager der Fall ist.  Die Funktion freePort ist zwar schon vorhanden, aber noch nicht in den entsprechenden Modulen berücksichtigt.  

In meinem Setup habe ich aktuell nur den NRF24 aktiv und in Nutzung. Der CMT ist zwar aktiviert, und somit die SPI Resource zugewiesen, aber nicht in Nutzung. Soweit funktioniert bei mir mein Setup.

**Reply by Manos1966** [reply]
Hi @skippermeister 
du hast die Frage beantwortet und am gleichen Tag die Anpassung programmiert? Ich bin beeindruckt! 🤩 
Die Idee kommt von der Tatsache, das die ESP32-S3 4 SPI haben. 
![SPI](https://github.com/user-attachments/assets/6965f8b4-0039-4a08-b967-00959a9009d4)

Und da die neuen Hoymiles Wechselrichter ausschließlich CMT2300A haben, werden neue Mitglieder nach einer Loesung fragen.
Vor allem, Leute die bereits eine grosse PV Anlage am Dach haben wo eine Batterie-Erweiterung (zu) viel Geld kostet. 😉

**Reply by skippermeister** [reply]
aber nur 2 sind verfügbar, die anderen beiden sind intern für den Flash Speicher und für PSRAM reserviert. Diese sollten auf keinen Fall verwendet werden.

**Reply by Manos1966** [reply]
You just ruined my evening... 😭 

Mist!

**Reply by Manos1966** [reply]
OK, after sleeping over it.... it does not seem so bad! 😄 

Welche Dateien hast du angepasst um den SPIPortManager zu realisieren?

Vorausgesetzt, ich erinnere mich richtig: 
Das Problem sind die Treiber des CMT2300. Sie kommen aus einer dritten Partei und haben einige "Einschraenkungen". 
Bis eine Verbesserung UPSTREAM gefunden wird, es ist so wie es ist und SPI2 wird vom CMT2300 zu stark beansprucht. 
Wenn das Huawei Modul angepasst wird um den jeweils freien SPI nutzen zu koennen, haetten wir vielleicht die Loesung. 🤞 

Ich baue gerade einen Huawei zusammen (reine wissenschaftliche Neugier)
![Huawei 4850G2](https://github.com/user-attachments/assets/d6058aec-7ece-4278-8df6-cacbc69e3141)

**Reply by Manos1966** [reply]
@skippermeister Vielen Dank 🥇

**Reply by skippermeister** [reply]
@Manos1966 
eine neues Release mit Korrekturen im SPIPortManager habe ich soeben hochgeladen.

**Comment by Manos1966** [comment]
Hallo @skippermeister 

ich habe gerade die Release 2024.08.01 opendtu-onbattery-generic_esp32_4mb_no_ota.factory.bin geflascht.

Ich bekomme aber den CMT2300A nicht aktiviert!
Ich habe zwei Platinen probiert, das gleiche Ergebnis
PIN Mapping ist korrekt.
![no CMT](https://github.com/user-attachments/assets/2ecdfa81-73b3-491e-bc0e-8930c77d9bf6)

Die gleiche Platine, mit der gleichen Pin-out Datei, und die Helegeerbe Release https://github.com/helgeerbe/OpenDTU-OnBattery/releases/tag/2024.06.03
hat kein Problem den CMT2300A zu erkennen:
![mit CMT](https://github.com/user-attachments/assets/b761d68b-01c1-4b06-b19b-5316353c0869)

**Reply by skippermeister** [reply]
@Manos1966 : Schick mir mal dein Pinmapping.

**Reply by skippermeister** [reply]
Es kann auch nicht funktionieren, schau Dir platformio.ini die Section [env:generic_esp32_4mb_no_ota] an. Das opendtu-onbattery-generic_esp32_4mb_no_ota.factory.bin hat für den Compiler Lauf den Schalter "-DUSE_RADIO_CMT=1 nicht gesetzt. Es ist also gar nicht das CMT Software Modul eingebunden. Soll ich Dir eine Version kompilieren oder kannst Du das selbst. Ach ich kompiliere schnell und lade die bin Datei hoch. Bekommt einen Namen opendtu-onbattery-generic_esp32_4mb_nrf_cmt_no_ota.factory.bin

**Comment by skippermeister** [comment]
ist jetzt hochgeladen, versuch es mal.

**Reply by Manos1966** [reply, marked as answer]
> Es kann auch nicht funktionieren, schau Dir platformio.ini die Section

Dan haette ich Lange versuchen koennen 😆😆😆😆

Vielen Dank @skippermeister fuer die hochgeladene Version.

Ich wollte testen, weil ich Probleme mit Andreas Version hatte... es hat sich herausgestellt es lag nicht an meiner HW.
Jetzt funktioniert auch seine Version (ich muss sie noch mit einem HMS Wechselrichter vollständig testen)
https://github.com/helgeerbe/OpenDTU-OnBattery/pull/1144#issuecomment-2265415130

Ich kann es nicht fassen, dass binnen 2 Wochen du und Andreas, das leidige Thema "Huawei+CMT2300A" so schnell erledigt habt! Vielen herzlichen Dank dafuer ❤️
